drake/solvers/NloptSolver.cpp

```diff
@@ -0,0 +1,291 @@
+  std::vector<double> xupp(nx, std::numeric_limits<double>::infinity());
+
+  for (auto const& binding : prog.getBoundingBoxConstraints()) {
+    auto const& c = binding.getConstraint();
+    for (const DecisionVariableView& v : binding.getVariableList()) {
+      auto const lb = c->getLowerBound(), ub = c->getUpperBound();
+      for (int k = 0; k < v.size(); k++) {
+        const int idx = v.index() + k;
+        xlow[idx] = std::max(lb(k), xlow[idx]);
+        xupp[idx] = std::min(ub(k), xupp[idx]);
+        if (x[idx] < xlow[idx]) { x[idx] = xlow[idx]; }
+        if (x[idx] > xupp[idx]) { x[idx] = xupp[idx]; }
+      }
+    }
+  }
+
+  opt.set_lower_bounds(xlow);
+  opt.set_upper_bounds(xupp);
+
+  opt.set_min_objective(EvaluateCosts, &prog);
+
+  // TODO sam.creasey All hardcoded tolerances in this function should
+  // be made configurable when #1879 is fixed.
+
+  std::list<WrappedConstraint> wrapped_list;
+
+  // TODO sam.creasey Missing test coverage for generic constraints
+  // with >1 output.
+  for (auto& c : prog.getGenericConstraints()) {
+    WrappedConstraint wrapped = { c.getConstraint().get(),
+                                  &c.getVariableList() };
+    wrapped_list.push_back(wrapped);
+    std::vector<double> tol(c.getConstraint()->getNumConstraints(), 1e-8);
+    opt.add_inequality_mconstraint(
+        EvaluateVectorConstraint, &wrapped_list.back(), tol);
+  }
+
+  // sam.creasey: The initial implementation of this code used
+  // add_equality_mconstraint to handle constraints with multiple
+  // outputs as a vector constraint.  This did not seem to work.  The
+  // version below breaks out the problem into multiple constraints.
+  std::list<LinearEqualityConstraint> equalities;
+  for (auto& c : prog.getLinearEqualityConstraints()) {
+    const size_t num_constraints = c.getConstraint()->getNumConstraints();
+    const auto& A = c.getConstraint()->getMatrix();
+    const auto& b = c.getConstraint()->getLowerBound();
+    for (size_t i = 0; i < num_constraints; i++) {
+      equalities.push_back(LinearEqualityConstraint(A.row(i), b.row(i)));
+      WrappedConstraint wrapped = { &equalities.back(),
+                                    &c.getVariableList() };
+      wrapped_list.push_back(wrapped);
+      opt.add_equality_constraint(
+          EvaluateScalarConstraint, &wrapped_list.back(), 1e-8);
+    }
+  }
+
+  // TODO sam.creasey Missing test coverage for generic constraints
+  // with >1 output.
+  for (auto& c : prog.getLinearConstraints()) {
+    WrappedConstraint wrapped = { c.getConstraint().get(),
+                                  &c.getVariableList() };
+    wrapped_list.push_back(wrapped);
+    std::vector<double> tol(c.getConstraint()->getNumConstraints(), 1e-8);
+    opt.add_inequality_mconstraint(
+        EvaluateVectorConstraint, &wrapped_list.back(), tol);
+  }
+
+  opt.set_xtol_rel(1e-6);
+
+  double minf = 0;
+  opt.optimize(x, minf);
```

Comment:
It seems to reliably throw in the event of failure.  I was going to skip checking the result unless I had some particular reason to handle it.  The alternative would be to eat the exception at this level and return false.